src/appfl/algorithm/server_fed_buffer.py

```diff
@@ -0,0 +1,90 @@
+import logging
+import copy
+from collections import OrderedDict
+from .server_federated import FedServer
+from ..misc import *
+
+logger = logging.getLogger(__name__)
+
+class ServerFedBuffer(FedServer):
+    """ Implement FedBuffer algorithm
+        Asynchronous Federated Optimization: https://arxiv.org/abs/2106.06639
+    
+    Agruments:
+        weights: weight for each client
+        model (nn.Module): PyTorch model
+        loss_fn (nn.Module): loss function
+        num_clients (int): number of clients
+        device (str): server's device for running evaluation  
+    """
+    def __init__(self, weights, model, loss_fn, num_clients, device, **kwargs):
+        self.counter = 0 
+        self.global_step = 0
+        weights = [1.0 / num_clients for _ in range(num_clients)] if weights is None else weights
+        super(ServerFedBuffer, self).__init__(weights, model, loss_fn, num_clients, device, **kwargs)
+        self.staleness = self.__staleness_func_factory(
+            stalness_func_name= self.staleness_func['name'],
+            **self.staleness_func['args']
+        )
+        for name, _ in self.model.named_parameters():
+            self.pseudo_grad[name] = torch.zeros_like(self.model.state_dict()[name])
+
+    def update_gradient(self, local_gradient: dict, init_step: int, client_idx: int):
+        for name, _ in self.model.named_parameters():
+            alpha_t = self.alpha * self.staleness(self.global_step - init_step)
+            self.pseudo_grad[name] += local_gradient[name]  * self.weights[client_idx] * alpha_t
+        self.counter += 1
+
+    def update(self, local_gradient: dict, init_step: int, client_idx: int):  
+        self.update_gradient(local_gradient, init_step, client_idx)
+        if self.counter == self.K:
```

Comment:
Where is `self.K` set?

Reply:
In the example file `examples/mnist_async_mpi`  line 166, we have `cfg.fed.args.K = args.K`, which puts K into the fed.args. 
And when we instantiate the server, we use the following, where K is passed by `**cfg.fed.args`
`server = eval(cfg.fed.servername)(
        weights, 
        copy.deepcopy(model), 
        loss_fn, num_clients, 
        device,
        **cfg.fed.args
    )
`
For ServerFedBuffer, it inherits FedServer, which deals with those keyword arguments by calling `self.__dict__.update(kwargs)`.